### PR TITLE
[ENH] add class weight to InceptionTime and capabability tag

### PR DESCRIPTION
**Pull request by Virgile-Foussereau**
#### Reference Issues/PRs

Add #8458 

#### What does this implement/fix? Explain your changes.

Add class weight capabability to InceptionTime and relevant tag

#### Does your contribution introduce a new dependency? If yes, which one?

No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great, thanks for the addition!

Could you kindly also add:

* the new tag in the `BaseClassifier` with a default of `False`
* the new tag in the extension template `extension_templates/classifier.py`
* a test in `TestAllClassifiers` for all classifiers with `class_weight`? Check the tag, and `return None` if the capability does not exist. If it does, use `set_params` to set `class_weight` based on the data passed.
    * Let us know if you need help with this!

**Comment by virgile-fsr** [comment]
> Great, thanks for the addition!
> 
> Code formatting is failing, see here how to set this up in automated fashion: https://www.sktime.net/en/stable/developer_guide/coding_standards.html
> 
> Could you kindly also add:
> 
> * the new tag in the `BaseClassifier` with a default of `False`
> * the new tag in the extension template `extension_templates/classifier.py`
> * a test in `TestAllClassifiers` for all classifiers with `class_weight`? Check the tag, and `return None` if the capability does not exist. If it does, use `set_params` to set `class_weight` based on the data passed.
>   
>   * Let us know if you need help with this!

Format should good now, I also added the tag and a basic test. Let me know if that works for you.

File: CONTRIBUTORS.md [hunk 1]
```diff
@@ -444,6 +444,7 @@ Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/d
...
       <td align="center" valign="top" width="11.11%"><a href="https://github.com/vollmersj"><img src="https://avatars2.githubusercontent.com/u/12613127?v=4?s=100" width="100px;" alt="vollmersj"/><br /><sub><b>vollmersj</b></sub></a><br /><a href="https://github.com/sktime/sktime/commits?author=vollmersj" title="Documentation">📖</a></td>
       <td align="center" valign="top" width="11.11%"><a href="https://github.com/xiaobenbenecho"><img src="https://avatars.githubusercontent.com/u/17461849?v=4?s=100" width="100px;" alt="xiaobenbenecho"/><br /><sub><b>xiaobenbenecho</b></sub></a><br /><a href="https://github.com/sktime/sktime/commits?author=xiaobenbenecho" title="Code">💻</a></td>
+      <td align="center" valign="top" width="11.11%"><a href="https://github.com/Virgile-Fousserequ"><img src="https://avatars.githubusercontent.com/u/99748194?v=4?s=100" width="100px;" alt="Virgile Foussereau"/><br /><sub><b>Virgile-Foussereau</b></sub></a><br /><a href="https://github.com/sktime/sktime/commits?author=Virgile-Foussereau" title="Code">💻</a></td>
```
Comment (fkiraly) [inline CONTRIBUTORS.md]: this is the wrong place to add yourself - please use `.all-contributorsrc`, the `contributors.md` is generated from that (and gets overwritten)

File: sktime/registry/_tags.py [hunk 2]
```diff
@@ -1041,6 +1041,40 @@ class capability__predict_proba(_BaseTag):
...
+    - String name: ``"capability:class_weight"``
+    - Public capability tag
+    - Values: boolean, ``True`` / ``False``
+    - Example: ``True``
+    - Default: ``False``
+
+    This tag applies to classifiers only.
+
+    If the tag is ``True``, the classifier supports class weighting functionality,
+    typically through a ``class_weight`` parameter that allows users to assign
```
Comment (fkiraly) [inline sktime/registry/_tags.py]: remove "typically" - the test enforces that specific name (and that is a good idea)

File: sktime/classification/deep_learning/inceptiontime.py [hunk 3]
```diff
@@ -104,6 +108,7 @@ def __init__(
         verbose=False,
         loss="categorical_crossentropy",
         metrics=None,
+        class_weight=None
```
Comment (fkiraly) [inline sktime/classification/deep_learning/inceptiontime.py]: please add this to the docstring

File: sktime/registry/_tags.py [hunk 4]
```diff
@@ -1041,6 +1041,40 @@ class capability__predict_proba(_BaseTag):
...
+    different weights to different classes during training. This is commonly used
+    to handle imbalanced datasets where some classes are underrepresented.
+
+    Class weights can usually be provided as:
+    - A dictionary mapping class labels to weights
```
Comment (fkiraly) [inline sktime/registry/_tags.py:1200]: is "balanced" a required feature?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great! Small change requests:

* please also update the docstring with the new parameter
* contributors should be added to another file

File: sktime/registry/_tags.py (hunk 4 above)
Comment (Virgile-Foussereau) [reply sktime/registry/_tags.py:1200]: Removed

File: sktime/classification/deep_learning/inceptiontime.py (hunk 3 above)
Comment (Virgile-Foussereau) [reply sktime/classification/deep_learning/inceptiontime.py]: Done

File: sktime/registry/_tags.py (hunk 2 above)
Comment (Virgile-Foussereau) [reply sktime/registry/_tags.py]: Done

File: CONTRIBUTORS.md (hunk 1 above)
Comment (Virgile-Foussereau) [reply CONTRIBUTORS.md]: Thanks

File: sktime/classification/deep_learning/inceptiontime.py [hunk 5]
```diff
@@ -48,6 +48,7 @@ class InceptionTimeClassifier(BaseDeepClassifier):
         whether to print runtime information
     loss: str, default="categorical_crossentropy"
     metrics: optional
+    class_weight: dict, optional, default=None
```
Comment (fkiraly) [inline sktime/classification/deep_learning/inceptiontime.py:52]: this is not enough, it should explain the expected format. Surely not any `dict` can be passed, and a user has no way to guess.

Comment (Virgile-Foussereau) [reply sktime/classification/deep_learning/inceptiontime.py:52]: It is now more detailed.

Comment (fkiraly) [reply sktime/classification/deep_learning/inceptiontime.py:52]: Thanks! One more clarification needed: what happens if class labels are not integers?

Comment (Virgile-Foussereau) [reply sktime/classification/deep_learning/inceptiontime.py:52]: Tensorflow specifies they should be integers: https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit

Comment (fkiraly) [reply sktime/classification/deep_learning/inceptiontime.py:52]: sure, but

(a) this is tensorflow, and
(b) in sktime, it is not clear how they would map onto arbitrary classification labels.

What happens in the case where the labels are not numbers?

**Comment by fkiraly** [comment]
code formatting fails, here is the guide:
https://www.sktime.net/en/stable/developer_guide/coding_standards.html

**Comment by fkiraly** [comment]
following up on this - I would like to include this in the next release, it is a really nice feature!

The docstring needs to say though how this is applied to classes that are not integers. Kindly note that it is possible to pass non-integer class vectors to `sktime` classifiers, and the user will not be aware of what happens under the hood!

File: sktime/registry/_tags.py (hunk 4 above)
Comment (fkiraly) [reply sktime/registry/_tags.py:1200]: Thanks!

Btw, pleaes ensure there are newlines directly before and after bullet point lists, or they will not render properly.

File: sktime/classification/deep_learning/inceptiontime.py (hunk 5 above)
Comment (Virgile-Foussereau) [reply sktime/classification/deep_learning/inceptiontime.py:52]: Ok I now use the label encoder to handle arbitrary labels

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great, I think this will take care of non-integer labels!

Some small change requests:

* code formatting is failing
* docstring for the new tag should have spaces around bullet point lists to render correctly
* can we please not raise an error if the `class_weight` dict does contain more classes than the data? Otherwise the estimator will not be stable with respect to subsetting the data, which can cause unexpected failures in boosting, bagging, or other ensembles which row-subset

**Comment by fkiraly** [comment]
(started the tests - code formatting issues)

**Review by fkiraly — APPROVED** [review]
Tests pass!
